### ISSUE TITLE
Is there an F# outline view?

### ISSUE THREAD

**Issue by Kurren123** [dotnet/fsharp#3947]
For large F# files, is there a way to see (and navigate to) all the modules and functions in that file in the order they are declared? The dropdown boxes at the top only show them in alphabetical order.

**Comment by vasily-kirichenko** [comment]
It's not implemented and there are not plans to add it.

**Comment by enricosada** [comment]
as a note, if @Kurren123 mean something like 

![image](https://user-images.githubusercontent.com/147243/32841357-562e5e90-ca1a-11e7-82ea-418b8eeea0e8.png)

the code is https://github.com/ionide/ionide-vscode-fsharp/blob/310a03e1d03db3c6655cc438154d5455e9f37336/src/Components/CodeOutline.fs (so can be followed until FCS)

**Comment by vasily-kirichenko** [comment]
Structure is available of course. The problem is to implement the VS API.

**Comment by cartermp** [comment]
An alternative would be to see if Navigation Bars could be modified to show things in the order of their definitions.  We'd have to see if this is something that we can support, or if it would require a feature request from the underlying navigation bar infrastructure.  This would also have to be controlled by a toggle in the options.  For every person who would want to see things in top-down order in the nav bars, another would likely want to see them in alphabetical order.

There's also a question of where this would be in the VS UI - a separate window would be quite a bit of work, and it means lots of working with and maintaining code that interacts directly with the VS API 😱.

Implementing this as a toggle view in Solution Explorer would likely be the best option, but it presents its own problems:

* This would break solution search as it is today, so that would need work done to support toggle views like this
* MPF (which the current F# project system builds atop) most likely wouldn't be updated to add any support that we'd need, given that it's an old codebase that people have moved on from
* CPS (which the new .NET project system that F# uses for .NET Core SDK-based projects) could be modified to support what this would need, but I suspect it would be a lot of work, and the team frankly has bigger issues to tackle right now

In short, this is a big feature request.  The stuff needed in our own language service is easy enough to implement, but it's (unfortunately) not where the bulk of the work happens.

**Comment by dsyme** [comment]
Closing this as it is a new VS view that is not going to be done F#-specific
